src/HttpClient/HttpClient.php

```diff
@@ -0,0 +1,158 @@
+        curl_setopt($curlHandle, \CURLOPT_HEADER, true);
+        /**
+         * @TODO(michi) make this configurable
+         */
+        curl_setopt($curlHandle, \CURLOPT_SSL_VERIFYPEER, true);
+        /**
+         * @TODO(michi) make this configurable
+         *
+         * If we add support for CURL_HTTP_VERSION_2_0, we need
+         * case-insensitive header handling, as HTTP 2.0 headers
+         * are all lowercase.
+         */
+        curl_setopt($curlHandle, \CURLOPT_HTTP_VERSION, \CURL_HTTP_VERSION_1_1);
+
+        $httpProxy = $this->options->getHttpProxy();
+        if (null !== $httpProxy) {
+            curl_setopt($curlHandle, \CURLOPT_PROXY, $httpProxy);
+        }
+
+        $httpProxyAuthentication = $this->options->getHttpProxyAuthentication();
+        if (null !== $httpProxyAuthentication) {
+            curl_setopt($curlHandle, \CURLOPT_PROXYUSERPWD, $httpProxyAuthentication);
+        }
+
+        /**
+         * @TODO(michi) add request compression (gzip/brotli) depending on availiable extensions
+         */
+        $body = curl_exec($curlHandle);
+
+        if (false === $body) {
+            $errorCode = curl_errno($curlHandle);
+            $error = curl_error($curlHandle);
+            curl_close($curlHandle);
+
+            $message = 'cURL Error (' . $errorCode . ') ' . $error;
+
+            return new Response(0, [], $message);
+        }
+
+        $statusCode = curl_getinfo($curlHandle, \CURLINFO_HTTP_CODE);
+        $headerSize = curl_getinfo($curlHandle, \CURLINFO_HEADER_SIZE);
+        $headers = $this->getResponseHeaders($headerSize, (string) $body);
+
+        curl_close($curlHandle);
+
+        return new Response($statusCode, $headers, '');
+    }
+
+    /**
+     * @return string[]
+     */
+    protected function getRequestHeaders(): array
+    {
+        $headers = [
+            'Content-Type' => 'application/x-sentry-envelope',
+        ];
+
+        $dsn = $this->options->getDsn();
+        if (null === $dsn) {
+            return $headers;
+        }
+
+        $data = [
+            'sentry_version' => Client::PROTOCOL_VERSION,
+            'sentry_client' => $this->sdkIdentifier . '/' . $this->sdkVersion,
+            'sentry_key' => $dsn->getPublicKey(),
+        ];
+
+        if (null !== $dsn->getSecretKey()) {
+            $data['sentry_secret'] = $dsn->getSecretKey();
+        }
+
+        $authHeader = [];
+        foreach ($data as $headerKey => $headerValue) {
+            $authHeader[] = $headerKey . '=' . $headerValue;
+        }
+
+        return array_merge($headers, [
+            'X-Sentry-Auth' => 'Sentry ' . implode(', ', $authHeader),
+        ]);
+    }
+
+    /**
+     * @TODO(michi) This might need a bit more love,
+     * but we only really care about X-Sentry-Rate-Limits and Retry-After
+     *
+     * @return string[]
+     */
+    protected function getResponseHeaders(?int $headerSize, string $body): array
```

Comment:
This is a BIG hot spot: making some mistakes here would open the door for CVEs.

Reply:
Please expand on this. The only two header values currently used are `X-Sentry-Rate-Limits` and `Retry-After` and I fail to see how this could lead to a vulnerability.

Reply:
Just today I got a notification about a Guzzle CVE that hit them a few months ago with a 7.5 score about headers: https://security.snyk.io/vuln/SNYK-PHP-GUZZLEHTTPPSR7-5426094

Handling headers could open us to CVE reports.

Reply:
Yep, but I would argue that this heavily depends on the context of how the header values are used. We could limit the parsing on the mentioned two headers. In the context of Guzzle, I agree that this is much more critical than in our case.

Most SDKs will talk to a somewhat "trusted" server, be it Sentry SaaS or self-hosted, and even if someone sneaks in some tampered headers, the worst I can see right now is that the rate-limiting won't work.

Reply:
That's not true. My was just an example of what security implications we would be opening to if we start writing our own HTTP client.

That specific vulnerability allows the attacker to slip through a custom, malicious body in the response, but I could easily imagine a similar vulnerability in the opposite direction; so you would have the SDK as a vector attack toward our end user or toward Sentry.io or a self-hosted server.

And I'm just limiting myself to the headers. The attack surface may be reduced due to the specific usage of the client (for Sentry only) but it's still there, and it comprises headers, body, format/compression in both requests and responses.